### ISSUE TITLE
Elsa takes down your application if it can't connect to Kafka

### ISSUE THREAD

**Issue by davydog187** [bbalser/elsa#55]
Hello! Thanks for your great work on elsa.

I've been testing it out at work in a Docker Compose environment running with Zookeeper and Confluent. (I'll include the `docker-compose.yml` below). I started out by creating an Elixir Kafka producer using Elsa, but I had to move away from it as it would crash the application. The reason for this is that elsa's supervision strategy forces the main supervisor to crash if it can't immediately connect to Kafka. This is problematic in scenarios like this Docker compose environment, where Kafka may have been started, but is not "ready" to start serving.

I ended up switching back to [Kaffe](https://github.com/spreedly/kaffe) which suits the needs of this small project for now. I wanted to create this issue to start a discussion around how this problem might be solved. I am happy to help :)

```yml
version: '2'
services:
  zookeeper:
    image: confluentinc/cp-zookeeper:5.4.0
    hostname: zookeeper
    container_name: zookeeper
    ports:
      - '32181:32181'
    environment:
      ZOOKEEPER_CLIENT_PORT: 32181
      ZOOKEEPER_TICK_TIME: 2000
    extra_hosts:
      - "moby:127.0.0.1"

  kafka:
    image: confluentinc/cp-server:5.4.0
    hostname: kafka
    container_name: kafka
    ports:
      - '9092:9092'
      - '29092:29092'
    depends_on:
      - zookeeper
    environment:
      KAFKA_BROKER_ID: 1
      KAFKA_ZOOKEEPER_CONNECT: zookeeper:32181
      KAFKA_LISTENER_SECURITY_PROTOCOL_MAP: PLAINTEXT:PLAINTEXT,PLAINTEXT_HOST:PLAINTEXT
      KAFKA_INTER_BROKER_LISTENER_NAME: PLAINTEXT
      KAFKA_ADVERTISED_LISTENERS: PLAINTEXT://kafka:29092,PLAINTEXT_HOST://localhost:9092
      KAFKA_AUTO_CREATE_TOPICS_ENABLE: "true"
      KAFKA_OFFSETS_TOPIC_REPLICATION_FACTOR: 1
      KAFKA_CONFLUENT_LICENSE_TOPIC_REPLICATION_FACTOR: 1
      KAFKA_METRIC_REPORTERS: io.confluent.metrics.reporter.ConfluentMetricsReporter
      KAFKA_OFFSETS_TOPIC_REPLICATION_FACTOR: 1
      CONFLUENT_METRICS_REPORTER_BOOTSTRAP_SERVERS: kafka:29092
      CONFLUENT_METRICS_REPORTER_ZOOKEEPER_CONNECT: zookeeper:32181
      CONFLUENT_METRICS_REPORTER_TOPIC_REPLICAS: 1
      CONFLUENT_METRICS_ENABLE: 'true'
      CONFLUENT_SUPPORT_CUSTOMER_ID: 'anonymous'
  producer:
    build: ./producer
    depends_on:
      - "kafka"
    ports:
      - "4000:4000"
    environment:
      KAFKA_HOSTNAME: "kafka"
      KAFKA_PORT: "29092"
```

**Comment by bbalser** [comment]
Thank you for bringing up this issue.  I will look into this as soon as I can, hopefully later today.

**Comment by jeffgrunewald** [comment]
If you're looking to continue testing in the meantime, you might try using the `divo` library (available on hex) and supply a healthcheck with your docker-compose file. Divo will block your application from starting (and any tests from running) until the container registers as "healthy" and is ready to serve responses.

**Comment by bbalser** [comment]
We have validated the problem and have identified the cause.  We will be working on a fix over the next couple of days.

**Comment by davydog187** [comment]
Thanks for the reply @bbalser and @jeffgrunewald. I know of Divo but haven't tried it here. Thank you for the suggestion!

**Comment by davydog187** [comment]
@bbalser @jeffgrunewald is there any work in progress you can share on this issue? I would like to contribute

To work around this issue, I am currently using Kaffe to publish, and Elsa to create topics on startup, which mitigates most of my pain, but I'm interested in just using Elsa

**Comment by jeffgrunewald** [comment]
@bbalser has been working on this issue specifically and from our last conversation yesterday I believe he’s pretty close to putting it out there but I’ll let him speak to that when he has a chance. 

We had discussed coupling a solution for this and https://github.com/bbalser/elsa/issues/58 since there’s some overlap between them but I think we settled on him releasing the fix for this issue (that happens to address the first concern of 58) and then I’ll pick it up and bring in the polling mechanism to dynamically respond to topic changes in a separate release.

**Comment by jeffgrunewald** [comment]
Just checking in that we've still been working on this and a fix should be out there soon. Unfortunately work has been very busy lately and only loosely related to Elsa. Stay tuned.

**Comment by jeffgrunewald** [comment]
FYI, @bbalser just released version 0.12 that should address this issue. Let us know if there are still any outlying problems that we missed.

**Comment by davydog187** [comment]
Will test. thanks @jeffgrunewald
